### PR TITLE
run setup_keychain on macOS environments only

### PR DESCRIPTION
**Pull request by seanreinhardtapps**
### Checklist
- [x] I've run `bundle exec rspec` from the root directory to see all new and existing tests pass
- [x] I've followed the _fastlane_ code style and run `bundle exec rubocop -a` to ensure the code style is valid
- [x] I've read the [Contribution Guidelines](https://github.com/fastlane/fastlane/blob/master/CONTRIBUTING.md)
- [x] I've updated the documentation if necessary.

### Motivation and Context
`setup_travis` fails on the Linux CI agents in the setup_keychain action, throwing an error and failing the task.
To run Travis CI lanes on Mac and Linux agents, `setup_travis` must be conditionally run.

Since that action only applies to a macOS environment, it could be updated to skip the setup_keychain step on non macOS agents.
🔑
### Description
Exit early out of `setup_keychain` in the `setup_ci` action on non mac-OS envoronments, logging a message that it was skipped.

### Testing Steps
- Run `setup_ci` on a macOS agent without any change in behavior.
- Run `setup_ci` on a linux agent and see that it no longer fails, and logs "Skipping Keychain setup on non-macOS CI Agent"

### CODE REVIEW

**Comment by seanreinhardtapps** [comment]
@googlebot I signed it!

**Comment by janpio** [comment]
Seems you will need some special handling in the tests as well, as they are run on Linux and Windows as well.

**Comment by mollyIV** [comment]
Hey @seanreinhardtapps , @janpio  👋 

The `setup_ci` lane has been extracted as a common piece of logic for specific CI providers. It makes maintenance and future code changes easier. [Here](https://github.com/fastlane/fastlane/pull/14346) is the full context.

According to the source code of a `setup_travis` lane, the action is intended to be ran only on `ios` and `mac` platforms:

https://github.com/fastlane/fastlane/blob/5734f464c121427e5505a632fe28fc944a141540/fastlane/lib/fastlane/actions/setup_travis.rb#L42-L44 

I had a look at `setup_ci` code and besides setting up the keychain and setting up the output paths only for a CircleCI provider, nothing is happening. 

https://github.com/fastlane/fastlane/blob/5734f464c121427e5505a632fe28fc944a141540/fastlane/lib/fastlane/actions/setup_ci.rb#L4-L16

Taking that into account, I am wondering if `setup_travis` should be supported on systems other than macOS right now 🤔 Maybe we could do it just in time (when adding new functionality to the `setup_travis` action that could be supported on the macOS, Linux and Windows), not just in case. 

@seanreinhardtapps could you please provide a bit more context, ideally a use-case, why and when you need to run `setup_travis` on an environment different than macOS?

**Comment by seanreinhardtapps** [comment]
@mollyIV, @janpio thanks for the reviews on this idea.
It's not my intent to actually run `setup_travis` on a linux agent.  My thought was that the `setup_travis` action could skip running on a linux agent, like it already does on a non-CI environment.  

When I looked into the implementation, I thought it would make more sense to skip the `setup_keychain` portion since thats what fails outside of a Mac, and that expands it beyond just Travis CI.

I understand it's not usual to use a linux agent for iOS tests.  I'm running a multi-stage Travis config where the last stage runs a docker container on the agent and starts XCUITests on a device farm.  Thought it could be just a ruby script, it's scripted into a lane along with all of our other tasks.  We had `setup_travis` at the top of our Fastfile before the need for the linux agent.  Now its run conditionally which is fine, but I thought it would be an improvement if it did so automatically.

File: fastlane/lib/fastlane/actions/setup_ci.rb [hunk 1]
```diff
@@ -29,6 +29,11 @@ def self.setup_keychain
           return
         end
 
+        unless Helper.operating_system == 'macOS'
```
Comment (joshdholtz) [inline fastlane/lib/fastlane/actions/setup_ci.rb:32]:
```suggestion
        unless Helper.mac?
```

File: fastlane/spec/actions_specs/setup_ci_spec.rb [hunk 2]
```diff
@@ -1,24 +1,33 @@
 describe Fastlane do
   describe Fastlane::Actions::SetupCiAction do
+    is_macos_env = RUBY_PLATFORM.downcase.include?("darwin")
```
Comment (joshdholtz) [inline fastlane/spec/actions_specs/setup_ci_spec.rb:3]: I think we could probably rework this so the tests aren't dependent on the platform they are being run on 😇 

The following tests should be reworked like 👇 
```rb
describe "calls setup_keychain after setup_output_paths if :provider is set to circleci on Mac Agents" do
  it "on Mac" do
    allow(FastlaneCore::Helper).to receive(:mac?).and_return(true)

    # Message is asserted in reverse order, hence output of setup_output_paths is expected last
    expect(Fastlane::UI).to receive(:message).with("Enabling match readonly mode.")
    expect(Fastlane::UI).to receive(:message).with("Creating temporary keychain: \"fastlane_tmp_keychain\".")
    expect(Fastlane::UI).to receive(:message).with("Skipping Log Path setup as FL_OUTPUT_DIR is unset")
  end

  it "on linux" do
    allow(FastlaneCore::Helper).to receive(:mac?).and_return(false)

    expect(Fastlane::UI).to receive(:message).with("Skipping Log Path setup as FL_OUTPUT_DIR is unset")
    expect(Fastlane::UI).to receive(:message).with("Skipping Keychain setup on non-macOS CI Agent")
  end
end
```

File: fastlane/spec/actions_specs/setup_circle_ci_spec.rb [hunk 3]
```diff
@@ -2,6 +2,21 @@
   describe Fastlane::Actions::SetupCircleCiAction do
     describe "Setup CircleCi Integration" do
       let(:tmp_keychain_name) { "fastlane_tmp_keychain" }
+      is_macos_env = RUBY_PLATFORM.downcase.include?("darwin")
```
Comment (joshdholtz) [inline fastlane/spec/actions_specs/setup_circle_ci_spec.rb:5]: See above comment for making tests not dependent on platform being executed on ☝️

File: fastlane/spec/actions_specs/setup_travis_spec.rb [hunk 4]
```diff
@@ -2,6 +2,21 @@
   describe Fastlane::FastFile do
     describe "Setup Travis Integration" do
       let(:tmp_keychain_name) { "fastlane_tmp_keychain" }
+      is_macos_env = RUBY_PLATFORM.downcase.include?("darwin")
```
Comment (joshdholtz) [inline fastlane/spec/actions_specs/setup_travis_spec.rb:5]: See above comment for making tests not dependent on platform being executed on ☝️

File: fastlane/lib/fastlane/actions/setup_ci.rb (hunk 1 above)
Comment (joshdholtz) [reply fastlane/lib/fastlane/actions/setup_ci.rb:32]: I also think this should get moved up before the checking of `ENV["MATCH_KEYCHAIN_NAME"]`

**Review by joshdholtz — CHANGES_REQUESTED** [review]
I think the addition of platform checking before setting up keychain makes sense! I have a few suggestions on the code in `SetupCiAction` and on the tests 😊 But once those are fixed I think this should be 👍

**Comment by mollyIV** [comment]
> I think the addition of platform checking before setting up keychain makes sense! 

👍 👍

Let's also remember to change a supported platforms for a `setup_travis` action 😊 

https://github.com/fastlane/fastlane/blob/5734f464c121427e5505a632fe28fc944a141540/fastlane/lib/fastlane/actions/setup_travis.rb#L42-L44

**Comment by joshdholtz** [comment]
@seanreinhardtapps BTW, let me know if you want any help making changes to those tests! I'd be happy to hop in and help 😊

**Comment by seanreinhardtapps** [comment]
Thanks for the reviews!  PR should be ready for another look.

**Review by joshdholtz — APPROVED** [review]
This is great! Thank you for making those changes 😊 I will get this out in a release later today for you 💪

**Pull request merged**